testutils: improve TestCluster{Interface,Factory} API

irfansharif opened the pull request:
The primitives exposed by the testutils package essentially boiled down
to "instantiate and start". This applied to TestServer, and working
backwards from there, it also applied to TestCluster. We didn't have a
way then to grab a handle on a Test{Cluster,Server} that hadn't fully
started yet.

Having such an ability is useful to have when testing
initialization/bootstrap behavior. There (in future PRs) we'd like to be
able to create an arbitrarily connected TestCluster and ensure that
bootstrap works as expected. Today to do this sort of thing we'd reach
for the roachtest infrastructure (like we do in
`acceptance/cluster-init`), but there's no reason we shouldn't be able
to do the same using the primitives provided by pkg/testutils.

To this end we separate "instantiation" and "start" into separate code
paths, and maintain the "instantiate and start" APIs but stub them out
to use these new APIs instead. Look towards pkg/testutils for the main
diff.

---

Specific code changes this commit brings about, for ease of review:
```
- Renames serverutils.StartTestCluster to serverutils.StartNewTestCluster
- TestServer.Start no longer takes base.TestServerArgs (they're still needed
  when creating the TestServer, avoiding the need to pass them again)
- Renames TestCluster.AddServer to TestCluster.AddAndStartServer
- (re-)Introduces TestCluster.AddServer to only add a server without starting
  it.
- Introduces TestClusterFactory.NewTestCluster to only create a new test
  cluster. This is accessible via serverutils.NewTestCluster.
- Deletes TestClusterFactory.StartTestCluster (make it more like a real
  factory).
- Introduces TestClusterInterface.Start to start clusters created using
  NewTestCluster.
- Introduces serverutils.NewServer to only create a new TestServer without
  starting it.
```

Release note: None
Release justification: non-production code changes

irfansharif committed:
testutils: improve TestCluster{Interface,Factory} API

The primitives exposed by the testutils package essentially boiled down
to "instantiate and start". This applied to TestServer, and working
backwards from there, it also applied to TestCluster. We didn't have a
way then to grab a handle on a Test{Cluster,Server} that hadn't fully
started yet.

Having such an ability is useful to have when testing
initialization/bootstrap behavior. There (in future PRs) we'd like to be
able to create an arbitrarily connected TestCluster and ensure that
bootstrap works as expected. Today to do this sort of thing we'd reach
for the roachtest infrastructure (like we do in
`acceptance/cluster-init`), but there's no reason we shouldn't be able
to do the same using the primitives provided by pkg/testutils.

To this end we separate "instantiation" and "start" into separate code
paths, and maintain the "instantiate and start" APIs but stub them out
to use these new APIs instead. Look towards pkg/testutils for the main
diff.

---

Specific code changes this commit brings about, for ease of review:

(1) Renames serverutils.StartTestCluster to serverutils.StartNewTestCluster
(2) TestServer.Start no longer takes base.TestServerArgs (they're still
    needed when creating the TestServer, avoiding the need to pass them
    again)
(3) Renames TestCluster.AddServer to TestCluster.AddAndStartServer
(4) (re-)Introduces TestCluster.AddServer to only add a server without
    starting it.
(5) Introduces TestClusterFactory.NewTestCluster to only create a new test
    cluster. This is accessible via serverutils.NewTestCluster.
(6) Deletes TestClusterFactory.StartTestCluster (make it more like a real
    factory).
(7) Introduces TestClusterInterface.Start to start clusters created using
    NewTestCluster.
(8) Introduces serverutils.NewServer to only create a new TestServer
    without starting it.
(9) Changes TestCluster.WaitForStores to TestCluster.WaitForNStores to
    have manual control over how many store descriptors the test cluster
    expects to find gossiped.

Release note: None
Release justification: non-production code changes

irfansharif commented:
@knz, I'm a bit lost with the CI failures. 

```
test_log_scope.go:107: can't use TestLogScope with secondary loggers active
```

I see you were in the area recently, as of https://github.com/cockroachdb/cockroach/pull/50506, but I'm not sure why the changes here (which is mostly benign code movement that had nothing to do with logging) are now failing because of it. These tests also pass locally (also tried with -v, -show-logs, etc.)

tbg reviewed (approved):
> Reviewed 46 of 46 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

knz commented:
> @knz, I'm a bit lost with the CI failures.

The error message tells you there's some secondary logger not being closed: a secondary logger is still running after _some_ test finishes, whereas the _next_ test fails because it can't open a log.Scope when there are secondary loggers active.

This means it's no use trying to run the tests marked as failed in CI individually - it's the test before that (which presumably succeeds) that matters.

Or you can run `make test` for the entire `kvserver` package. Which brings me to the next point:

Looking at the artifacts, I see this:

```
--- FAIL: TestClosedTimestampWorksWhenRequestsAreSentToNonLeaseHolders (8.17s)
test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestClosedTimestampWorksWhenRequestsAreSentToNonLeaseHolders887328149
test_log_scope.go:63: use -show-logs to present logs inline
stopper.go:99: leaked stopper, created at: [...]
```

leaking stoppers are never good, and it corroborates what I wrote above: there is some test server that does not get `Close()`'d properly.

I ran `make test` on `kvserver` on your branch. The test log is clear: the test `TestClosedTimestampWorksWhenRequestsAreSentToNonLeaseHolders` always fails with a leaked stopper, and then every test after that fails because there's a leaked secondary logger.

knz reviewed (commented):
> Reviewed 3 of 46 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53839)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)

___
*[pkg/testutils/testcluster/testcluster.go, line 368 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53839#-MGJ01RYBYpwhHqzYJZE:-MGJ01RZBcAkhPMY6PUU:b-xhd0y5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02b14c33dccc669e23594e8523336b4df9a2612f/pkg/testutils/testcluster/testcluster.go#L368)):*
> ```Go
> 		return nil, err
> 	}
> 	serverArgs.Stopper = stop.NewStopper()
> ```

@irfansharif I think that's your problem. 

There's a `Stopper` member in `TestCluster`, and that's what tests use to shut down the test cluster.
This additional Stopper here detaches the shutdown of the server added via `AddServer` from that of `tc.Stopper`.

This means any test that uses `AddServer` is flawed probably.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
Can you detail even more what exactly one would do with a `TestServer` that hasn't been started? And even more so with `TestCluster` that hasn't been started.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53839#-:-MGJWrsn6l-yp0WxFVdR:bgalewx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

irfansharif committed:
server: improve APIs for TestServer.{Start,New}

This commit is in the same spirit as the previous. We keep
TestServer.New limited to "initialization" and TestServer.Start to
actually "start" the server process. These methods already existed, we
just had to shuffle some initialization responsibility from the latter
to the former.

Release note: None
Release justification: non-production code change

irfansharif reviewed (commented):
> it's the test before that (which presumably succeeds) that matters.

oo, thanks for finding that, I should've clicked around on more of those failures. The problem was from usage of the AddAndStartCluster helper, which now informs TestCluster of the added server's stoppers.

> Can you detail even more what exactly one would do with a TestServer that hasn't been started? And even more so with TestCluster that hasn't been started.

I'm trying to add a test like this one: https://github.com/irfansharif/cockroach/blob/3fd156b8e3883da082941a518613b2111d5b2dae/pkg/server/connectivity_test.go#L35-L39. What it does is it sets up a possibly partially connected cluster, where some TestServers may never learn about the cluster ID, so they may never fully start up (consequently the TestCluster as a whole will never fully start up). This happens when the server, through its join flags, is pointing to other nodes that won't find out about bootstrap. The test wants to check that all the nodes we expect to join the cluster actually do, and the ones we don't, never do.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53839#-:-MGJfAOE16zDakdpiNIY:b-3op7cx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)



<!-- Sent from Reviewable.io -->

irfansharif commented:
bors r+

irfansharif commented:
bors r-

irfansharif commented:
bors single on

irfansharif commented:
bors r+

irfansharif commented:
Ugh, I don't understand the bors failure.

```
pkg/server/status_test.go:1680:17: undefined: serverutils.StartTestCluster
```

I can see in by branch that that was updated. It's not the right line.

irfansharif commented:
bors r-

irfansharif commented:
bors r+
